core/distributed/matrix.cpp

```diff
@@ -271,29 +270,40 @@ void Matrix<ValueType, LocalIndexType, GlobalIndexType>::read_distributed(
     array<local_index_type> local_col_idxs{exec};
     array<value_type> local_values{exec};
     array<local_index_type> non_local_row_idxs{exec};
-    array<local_index_type> non_local_col_idxs{exec};
+    array<global_index_type> global_non_local_col_idxs{exec};
     array<value_type> non_local_values{exec};
-    array<comm_index_type> recv_sizes_array{exec, num_parts};
 
-    // build local, non-local matrix data and communication structures
-    exec->run(matrix::make_build_local_nonlocal(
+    // separate input into local and non-local block
+    // The rows and columns of the local block are mapped into local indexing,
+    // as well as the rows of the non-local block. The columns of the non-local
+    // block are still in global indices.
+    exec->run(matrix::make_separate_local_nonlocal(
         data, make_temporary_clone(exec, row_partition).get(),
         make_temporary_clone(exec, col_partition).get(), local_part,
         local_row_idxs, local_col_idxs, local_values, non_local_row_idxs,
-        non_local_col_idxs, non_local_values, recv_gather_idxs_,
-        recv_sizes_array, non_local_to_global_));
+        global_non_local_col_idxs, non_local_values));
+
+    auto imap = index_map<local_index_type, global_index_type>(
+        exec, col_partition, comm.rank(), global_non_local_col_idxs);
```

Comment:
I'm quite confident that it will become necessary to store this map in the matrix class in the future. I'm not doing that for now, because with the distributed PGM, there are now ways to create a non-empty matrix without this read function. I don't know how to create the mapping in the PGM case, so there would be non-empty matrices, but with empty mappings. 
I didn't think this is suitable, so I don't store the mapping. But making it a member in the future would either require unnecessary copies or an interface break.
Any thoughts on this @upsj @pratikvn?

Reply:
I think it's a fundamental choice - do we want to consider the matrix storing the implicitly block-reordered matrix (every rank has a consecutive range of indices), or the original matrix? I'm fine with the latter, but it may complicate things as mappings global <--> local are more costly. But at the same time, I'm not sure that would ever be a performance bottleneck.

Reply:
Can you elaborate on the interface break?

Reply:
interface break: 
```
auto imap = mat->read_distributed(...);
```
will result in an error if `read_distributed` returns again void. The return type could be changed to `const index_map&` but this would result in copies.

Reply:
That's not entirely true - you can operate on the block-reordered indexing just fine, imagine computing an index_map after remapping each rank's indices to be consecutive. But that is still doing 90% of the work towards having a full global <--> local index map, so it's probably to better to do it fully.

Reply:
IMO, just from a properties perspective, I think it makes sense for the matrix class to have an index map. Any function that creates a distributed matrix, should then also be required to update its index_map. I understand that we dont do that for PGM, right now, but I think the interface choices should not be dependent on that. 

So, I think it should be okay to store an index_map in the matrix class. 

TBH, I am not sure why `read_distributed` should return an index_map. I think we should just store it inside the class and have a public getter to access it where necessary.

Reply:
I think the reordering question is a bit separate from the index-map question. As you said, you could still have an index map after reordering. In general, I think reordering can become quite complicated, see the fun image in petsc's user guide: https://petsc.org/release/manual/mat/#block-matrices

Reply:
I think for now I will just remove the return type. The index map is still helpful, since it simplifies the matrix kernels, while the information necessary for PGM is still available. Thus we could add the index map as a member at a later point if we want to.